v2.0/architecture/transaction-layer.md

```diff
@@ -132,27 +126,32 @@ CockroachDB efficiently supports the strongest ANSI transaction isolation level:
+- **Snapshot Isolation** _(Snapshot)_ transactions trade correctness in order to avoid retries when serializability violations are possible. This is achieved by always reading at an initial transaction timestamp, but allowing the transaction's commit timestamp to be pushed forward in the event of [transaction conflicts](#transaction-conflicts). Snapshot isolation cannot prevent an anomaly known as [write skew](https://en.wikipedia.org/wiki/Snapshot_isolation).
 
 ### Transaction Conflicts
 
-CockroachDB's transactions allow the following types of conflicts:
+CockroachDB's transactions allow the following types of conflicts that involve running into an intent:
 
 - **Write/Write**, where two `PENDING` transactions create Write Intents for the same key.
 - **Write/Read**, when a read encounters an existing Write Intent with a timestamp less than its own.
 
 To make this simpler to understand, we'll call the first transaction `TxnA` and the transaction that encounters its Write Intents `TxnB`.
 
-CockroachDB proceeds through the following steps until one of the transactions is aborted, has its timestamp moved, or enters the `PushTxnQueue`.
+CockroachDB proceeds through the following steps until one of the transactions is aborted, has its timestamp pushed, or enters the `PushTxnQueue`.
 
-1. If the transaction has an explicit priority set (i.e. `HIGH`, or `LOW`), the transaction with the lower priority is aborted.
+1. If the transaction has an explicit priority set (i.e. `HIGH`, or `LOW`), the transaction with the lower priority is aborted (in the writer/write case) or has its timestamp pushed (in the write/read case).
 
 2. `TxnB` tries to push `TxnA`'s timestamp forward.
 
     This succeeds only in the case that `TxnA` has snapshot isolation and `TxnB`'s operation is a read. In this case, the [write skew](https://en.wikipedia.org/wiki/Snapshot_isolation) anomaly occurs.
 
 3. `TxnB` enters the `PushTxnQueue` to wait for `TxnA` to complete.
 
+Additionally, the following types of conflicts that don't involve running into intents can arise:
+
+- **Write after read**, when a write with a lower timestamp encouters a later read. This is handled through the [Timestamp Cache](#timestamp-cache).
+- **Read within uncertainty window**, when a read encounters a value with a higher timestamp but it's ambiguous whether the value should be considered to be in the future or in the past of the transaction because of possible *clock skew*. This is handled by attempting to push the transaction's timestamp beyond the uncertain value (see [read refreshing](#read-refreshing)).
```

Comment:
s/attempting to push.../attempting to forward the transaction's timestamp just beyond the uncertain value's timestamp/ (I don't think push is right in this context).

You might also add: "If the transaction timestamp forwarding and refreshing is unsuccessful, then the transaction must be retried at the forwarded timestamp. Note that on successive retries, reads will never encounter uncertainty issues on any node which was previously visited."